Add RenderElement plugin to support appleseed's AOVs

usakhelo opened the pull request:
Hi @dictoon 
This is render element plugin that allows users to render appleseed AOVs.
Plugin basically used to store AOV type and output filename.
I could not find a way to copy AOV image from appliseed image to 3ds max render element bitmap, and for that reason at some stage render elements are turned off and files are saved with appleseed's function `project->get_frame()->write_main_and_aov_images();`

Thanks for review,
Sergo.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.h:
```diff
@@ -0,0 +1,134 @@
+#include "foundation/platform/windows.h"    // include before 3ds Max headers
+
+// 3ds Max headers.
+#include <renderelements.h>
```
Order alphabetically.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp:
```diff
@@ -37,6 +37,7 @@
 #include "appleseedrenderer/projectbuilder.h"
 #include "appleseedrenderer/renderercontroller.h"
 #include "appleseedrenderer/tilecallback.h"
+#include "appleseedrenderelement/appleseedrenderelement.h"
```
Order alphabetically.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp:
```diff
@@ -125,6 +129,10 @@ void* AppleseedRenderer::GetInterface(ULONG id)
     }
+    if (id == IRenderElementCompatible::IID)
+    {
+        return new AppleseedRECompatible();
```
Unline `AppleseedInteractiveRender` above, here we are returning a new instance each time. Is this intended?

usakhelo replied:
This class is only used to check element compatibility. But new instance each time is not needed, that's true. Also I don't know if it's get deleted anywhere later.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/appleseedrenderer.cpp:
```diff
@@ -240,6 +248,8 @@ int AppleseedRenderer::Open(
     int                     default_light_count,
     RendProgressCallback*   progress_cb)
 {
+    BroadcastNotification(NOTIFY_PRE_RENDER, static_cast<void*>(&rend_params));
```
The cast is redundant: in C++ all pointers can be implicitly converted to `void*`.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp:
```diff
@@ -1385,6 +1389,41 @@ namespace
+
+            IRenderElementMgr* re_manager = GetCOREInterface()->GetCurRenderElementMgr();
+            re_manager->SetDisplayElements(false);
+            for (size_t i = 0, e = re_manager->NumRenderElements(); i < e; ++i)
```
`NumRenderElements()` returns `int`; use `int` to iterate (and then you can remove a cast below).

dictoon commented on src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp:
```diff
@@ -1385,6 +1389,41 @@ namespace
+            re_manager->SetDisplayElements(false);
+            for (size_t i = 0, e = re_manager->NumRenderElements(); i < e; ++i)
+            {
+                const asr::AOVFactoryRegistrar factory_registrar;
```
Move the creation of the factory registrar out of the loop: creating a factory registrar is an extremely expensive operation as it will scan search paths for plugins, load DLLs, etc. In fact we might want to create the registrar once when the plugin is loaded rather than everytime a project is built.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp:
```diff
@@ -1385,6 +1389,41 @@ namespace
+                    AppleseedRenderElement* element = static_cast<AppleseedRenderElement*>(render_element);
+                    int aov_index = 0;
+                    element->GetParamBlock(0)->GetValueByName(L"aov_index", 0, aov_index, FOREVER);
+                    if (aov_index > 0 && aov_index <= factories.size())
```
Signed vs unsigned comparison here. I'm surprised you're not getting a warning. You will probably need to cast `factories.size()` to `int`.

dictoon commented on src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp:
```diff
@@ -1385,6 +1389,41 @@ namespace
+                        {
+                            PBBitmap* p_bitmap = nullptr;
+                            render_element->GetPBBitmap(p_bitmap);
+                            if (p_bitmap != nullptr)
```
Add braces around multi-line `if` body.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+    
+    std::vector<std::string> get_aov_names()
+    {
+        const asr::AOVFactoryRegistrar factory_registrar;
```
Same problem here. I'm not sure in which order you'll see my comments, but as I explained in another one, this is an extremely expensive operation.

usakhelo replied:
I moved it to the same place as `AppleseedRenderElementClassDesc g_appleseed_renderelement_classdesc;`

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+    {
+        const asr::AOVFactoryRegistrar factory_registrar;
+        auto factories = factory_registrar.get_factories();
+        auto aov_names = std::vector<std::string>();
```
Better:

    std::vector<std::string> aov_names;

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+        const asr::AOVFactoryRegistrar factory_registrar;
+        auto factories = factory_registrar.get_factories();
+        auto aov_names = std::vector<std::string>();
+        for (size_t i = 0, e = factories.size(); i < e; ++i)
```
Use range-based for loops whenever possible:

    for (const auto& factory : factories)
    {
        ...
    }

(You might need to adjust the type of `factory`.)

usakhelo replied:
Seems that `FactoryArrayType` can't be used with range loops, it says `error C3312: no callable 'begin' function found for type 'renderer::AOVFactoryRegistrar::FactoryArrayType'`

Should I do `auto factories = asf::array_vector<std::vector<asr::IAOVFactory*>>(g_aov_factory_registrar.get_factories());` ?

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+
+    void init_combo_box(IParamMap2* param_map)
+    {
+        HWND dlg_hwnd = 0;
```
I think that ultimately `HWND` is a pointer so `nullptr` would be more appropriate.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+    }
+}
+
+AppleseedRenderElement::~AppleseedRenderElement()
```
Remove the destructor if it does nothing?

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+    return REF_SUCCEED;
+}
+
+RefTargetHandle	AppleseedRenderElement::Clone(RemapDir& remap)
```
Remove tabulation.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,495 @@
+
+BOOL AppleseedRenderElement::IsEnabled() const
+{
+    int	is_on;
```
Remove tabulation.

usakhelo commented:
Hi @dictoon ,
I fixed your comments and also added a commit to fix the crash during the interactive rendering. Turned out during the interactive rendering RenderElementManager is not available, so I added a nullptr check. Note that this crash is not the same as the described in the issue #156.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -76,21 +78,20 @@ namespace
-        auto aov_names = std::vector<std::string>();
-        for (size_t i = 0, e = factories.size(); i < e; ++i)
-            aov_names.push_back(factories[i]->get_model_metadata().get("label"));
+        auto factories = asf::array_vector<std::vector<asr::IAOVFactory*>>(g_aov_factory_registrar.get_factories());
```
> Seems that FactoryArrayType can't be used with range loops, it says error C3312: no callable 'begin' function found for type 'renderer::AOVFactoryRegistrar::FactoryArrayType'

Oh ok, the code was fine then, sorry.

> Should I do auto factories = asf::array_vector<std::vector<asr::IAOVFactory*>>(g_aov_factory_registrar.get_factories()); ?

No please, that's a lot worse with allocations, constructions, destructions, copying... :)

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.h:
```diff
@@ -132,3 +137,4 @@ class AppleseedRenderElementClassDesc
 };
 
 extern AppleseedRenderElementClassDesc g_appleseed_renderelement_classdesc;
+extern renderer::AOVFactoryRegistrar g_aov_factory_registrar;
```
Add newline character.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,492 @@
+
+namespace
+{
+    
```
Remove blank line.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,492 @@
+            return;
+
+        SendMessage(GetDlgItem(dlg_hwnd, IDC_COMBO_AOV_NAME), CB_RESETCONTENT, 0, 0);
+        if (true)
```
Remove test?

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,492 @@
+    {
+      public:
+        void DeleteThis() override
+        {
```
Are we sure not doing anything here is correct?

usakhelo replied:
I thought it's fine since instance of this class is only created as a local object in anonymous namespace (line 176).

dictoon replied:
You're right, it's not an actual bug, however it's better to make the code as agnostic as possible to specific usage patterns, hence I suggest implementing this method (just `delete this` I presume?).

usakhelo replied:
I think usage pattern should be change in this case, otherwise I can't open rendering dialog box, it crashes the max.
I'll add `delete this` and assign dialog procedure in a way it is done in other parts of the plugin.

dictoon commented on src/appleseed-max-impl/appleseedrenderelement/appleseedrenderelement.cpp:
```diff
@@ -0,0 +1,492 @@
+            int output_index = param_map->GetParamBlock()->GetInt(ParamIdAOVIndex, 0, FOREVER);
+            DbgAssert(output_index >= 1);
+            output_index -= 1;
+            if (output_index < aov_names.size())
```
Signed/unsigned mismatch?
